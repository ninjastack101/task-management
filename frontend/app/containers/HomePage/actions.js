import {
  FETCH_TASKS,
  LOAD_TASKS,
  ADD_TASK,
  EDIT_TASK,
  DELETE_TASK
} from "./constants";

/**
 * Call Fetch request
 *
 * @param  {string}
 *
 * @return {object} An action object with a type of FETCH_TASKS
 */
export const fetchTasks = () => {
  return {
    type: FETCH_TASKS
  };
};

/**
 * FetchTasks response
 *
 * @param  {string} id, username
 *
 * @return {object} An action object with a type of SIGNUP_SUCCESS
 */
export const setTasks = tasks => {
  return {
    type: LOAD_TASKS,
    tasks
  };
};

/**
 * Add Task response
 *
 * @param  {string} title, descrption, deadline
 *
 * @return {object} An action object with a type of SIGNUP_SUCCESS
 */
export const addTask = (title, description, deadline) => ({
  type: ADD_TASK,
  title,
  description,
  deadline
});

/**
 * Edit task response
 *
 * @param  {string} id, title, description , deadline
 *
 * @return {object} An action object with a type of SIGNUP_SUCCESS
 */
export const editTask = (id, title, description, deadline) => ({
  type: EDIT_TASK,
  title,
  description,
  deadline,
  id
});

/**
 * Edit task response
 *
 * @param  {string} id
 *
 * @return {object} An action object with a type of SIGNUP_SUCCESS
 */
export const deleteTask = id => ({
  type: DELETE_TASK,
  id
});
