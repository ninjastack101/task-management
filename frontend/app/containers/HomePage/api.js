import { call } from "redux-saga/effects";
import request from "utils/request";
import config from "config";

export function* getTasksApi(userId) {
  const requestURL = `${config.apiRoot}/tasks?user_id=${userId}`;
  return yield call(request, requestURL);
}

export function* addTasksApi(data) {
  const requestURL = `${config.apiRoot}/tasks`;
  const option = {
    method: "POST",
    body: JSON.stringify(data)
  };
  return yield call(request, requestURL, option);
}

export function* editTasksApi(taskId, data) {
  const requestURL = `${config.apiRoot}/tasks/${taskId}`;
  const option = {
    method: "PUT",
    body: JSON.stringify(data)
  };
  return yield call(request, requestURL, option);
}

export function* deleteTasksApi(taskId) {
  const requestURL = `${config.apiRoot}/tasks/${taskId}`;
  const option = {
    method: "DELETE"
  };

  return yield call(request, requestURL, option);
}
