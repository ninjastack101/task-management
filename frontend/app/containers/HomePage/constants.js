export const FETCH_TASKS = 'boilerplate/Home/FETCH_TASKS';
export const LOAD_TASKS = 'boilerplate/Home/LOAD_TASKS';
export const ADD_TASK = 'boilerplate/Home/ADD_TASK';
export const EDIT_TASK = 'boilerplate/Home/EDIT_TASK';
export const DELETE_TASK = 'boilerplate/Home/DELETE_TASK';
