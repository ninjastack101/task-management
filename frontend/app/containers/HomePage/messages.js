import { defineMessages } from 'react-intl';

export const scope = 'boilerplate.containers.HomePage';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'Task Listing',
  },
  addTask: {
    id: `${scope}.addTask`,
    defaultMessage: 'Add Task',
  },
});
