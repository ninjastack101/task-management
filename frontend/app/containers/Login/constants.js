export const CALL_SIGNUP = "boilerplate/LOGIN/CALL_SIGNUP";
export const SIGNUP_SUCCESS = "boilerplate/LOGIN/SIGNUP_SUCCESS";
export const LOGOUT = "boilerplate/LOGIN/LOGOUT";
