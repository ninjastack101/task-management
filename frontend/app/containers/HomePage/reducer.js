import produce from 'immer';
import { LOAD_TASKS } from './constants';

// The initial state of the App
export const initialState = {
  tasks: [],
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_TASKS:
        draft.tasks = action.tasks;
        break;
    }
  });

export default homeReducer;
