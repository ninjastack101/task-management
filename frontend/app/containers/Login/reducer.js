import produce from "immer";
import { SIGNUP_SUCCESS, LOGOUT } from "./constants";

// The initial state of the App
export const initialState = {
  username: "",
  userId: ""
};

/* eslint-disable default-case, no-param-reassign */
const loginReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case SIGNUP_SUCCESS:
        draft.userId = action.id;
        draft.username = action.username;
        break;
      case LOGOUT:
        draft.userId = "";
        draft.username = "";
        break;
    }
  });

export default loginReducer;
