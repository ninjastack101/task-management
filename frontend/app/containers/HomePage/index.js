import React from "react";
import PropTypes from "prop-types";
import { Helmet } from "react-helmet";
import { FormattedMessage } from "react-intl";
import { connect } from "react-redux";
import { compose } from "redux";
import { createStructuredSelector } from "reselect";
import injectReducer from "utils/injectReducer";
import injectSaga from "utils/injectSaga";
import H2 from "components/H2";
import Grid from "@material-ui/core/Grid";
import Tooltip from "@material-ui/core/Tooltip";
import Button from "@material-ui/core/Button";
import { withStyles } from "@material-ui/core/styles";
import Modal from "components/Modal";
import io from "socket.io-client";
import config from "config";
import Table from "./table";
import messages from "./messages";
import { addTask, editTask, fetchTasks, deleteTask } from "./actions";
import { makeSelectTasks, makeSelectUserDetails } from "./selectors";
import reducer from "./reducer";
import saga from "./saga";

const socket = io(config.apiRoot);
const key = "home";

const styles = theme => ({
  paddingStyle: {
    padding: theme.spacing(3, 3, 2)
  }
});

class HomePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      shouldOpenModal: false,
      formData: {}
    };
    if (props.userDetails.userId) {
      socket.on(props.userDetails.userId, payload => {
        props.loadTasks();
      });
    }
  }

  componentWillUnmount() {
    socket.off(this.props.userDetails.userId);
    console.log("Socket Disconnected");
  }

  columns = [
    { id: "title", label: "Title", minWidth: 200 },
    { id: "description", label: "Description", minWidth: 100 },
    {
      id: "deadline",
      label: "Due Date",
      minWidth: 120,
      format: value => {
        if (!value) {
          return "";
        }
        let d = new Date(value);
        let date =
          parseInt(d.getMonth() + 1) +
          "-" +
          d.getDate() +
          "-" +
          d.getFullYear();

        return date;
      }
    }
  ];

  componentDidMount() {
    this.props.loadTasks();
  }

  toggleModal = flag => {
    this.setState({ shouldOpenModal: flag });
    this.setState({ formData: {} });
  };

  onSubmitForm = event => {
    event.preventDefault();
    const { id, title, description, deadline } = event.target.elements;
    if (!title.value.trim() || !description.value.trim()) {
      return;
    }
    if (id && id.value) {
      this.props.editTasks(
        id.value,
        title.value,
        description.value,
        deadline.value
      );
      this.setState({ formData: {}, shouldOpenModal: false });
    } else {
      this.props.addTasks(title.value, description.value, deadline.value);
      this.setState({ formData: {}, shouldOpenModal: false });
    }
  };

  updateTasks = task => {
    this.setState({ formData: task });
    this.setState({ shouldOpenModal: true });
  };

  render() {
    const { tasks, classes } = this.props;
    return (
      <article>
        <Helmet>
          <title>Home Page</title>
          <meta name="description" content="A React.js application homepage" />
        </Helmet>
        <H2 className={classes.paddingStyle}>
          <FormattedMessage {...messages.header} />
        </H2>
        <Grid
          item
          container
          alignItems="flex-end"
          direction="column"
          wrap="nowrap"
          spacing={10}
          className={classes.paddingStyle}
        >
          <Grid item>
            <Tooltip
              title={<FormattedMessage {...messages.addTask} />}
              placement="top"
            >
              <Button
                variant="contained"
                color="primary"
                onClick={() => this.toggleModal(true)}
              >
                <FormattedMessage {...messages.addTask} />
              </Button>
            </Tooltip>
          </Grid>
        </Grid>
        <div className={classes.paddingStyle}>
          <Table
            rows={tasks}
            columns={this.columns}
            handleClick={this.updateTasks}
            handleDelete={this.props.deleteTasks}
            className={classes.paddingStyle}
          />
        </div>
        {this.state.shouldOpenModal && (
          <Modal
            open={this.state.shouldOpenModal}
            onSubmit={this.onSubmitForm}
            toggle={this.toggleModal}
            task={this.state.formData}
          />
        )}
      </article>
    );
  }
}

HomePage.propTypes = {
  classes: PropTypes.object.isRequired,
  tasks: PropTypes.array,
  userDetails: PropTypes.object.isRequired,
  loadTasks: PropTypes.func,
  addTasks: PropTypes.func,
  editTasks: PropTypes.func,
  deleteTasks: PropTypes.func
};

const mapStateToProps = createStructuredSelector({
  tasks: makeSelectTasks(),
  userDetails: makeSelectUserDetails()
});

export function mapDispatchToProps(dispatch) {
  return {
    loadTasks: () => {
      dispatch(fetchTasks());
    },
    editTasks: (id, name, description, deadline) => {
      dispatch(editTask(id, name, description, deadline));
    },
    addTasks: (title, description, deadline) => {
      dispatch(addTask(title, description, deadline));
    },
    deleteTasks: id => {
      dispatch(deleteTask(id));
    }
  };
}

const withSaga = injectSaga({ key, saga });
const withReducer = injectReducer({ key, reducer });

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default compose(
  withStyles(styles, { name: "Home" }),
  withSaga,
  withReducer,
  withConnect
)(HomePage);
