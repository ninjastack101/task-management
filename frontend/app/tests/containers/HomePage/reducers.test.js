import reducer from '../../../containers/HomePage/reducer';
import { LOAD_TASKS } from '../../../containers/HomePage/constants';

describe('Fetch home reducer', () => {

  let initialState = {
    tasks: [],
  };

  it('Should return the initial state', (done) => {
    expect(reducer(undefined, {})).toEqual(initialState);
    done();
  });

  it('should handle FETCH_PRODUCTS_LIST_REQUEST', (done) => {

    const startAction = {
      type: LOAD_TASKS,
      tasks: {
        0: "somestring"
      }
    };
    expect(reducer(initialState, startAction)).toMatchSnapshot();
    done();
  });
});
