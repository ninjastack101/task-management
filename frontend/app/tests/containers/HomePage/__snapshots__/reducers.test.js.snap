// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch home reducer should handle FETCH_PRODUCTS_LIST_REQUEST 1`] = `
Object {
  "tasks": Object {
    "0": "somestring",
  },
}
`;
